File: redux/src/main/java/com/cyworks/redux/LogicPage.java
package com.cyworks.redux;

import android.os.Bundle;
import android.os.Looper;
import android.support.annotation.CallSuper;
import android.support.annotation.NonNull;

import com.tencent.redux.action.Action;
import com.tencent.redux.action.InnerActions;
import com.tencent.redux.beans.InterceptorBean;
import com.tencent.redux.dependant.DependentCollect;
import com.tencent.redux.dependant.ExtraDependants;
import com.tencent.redux.dispatcher.Dispatch;
import com.tencent.redux.effect.Effect;
import com.tencent.redux.effect.EffectCollect;
import com.tencent.redux.interceptor.Interceptor;
import com.tencent.redux.interceptor.InterceptorCollect;
import com.tencent.redux.interceptor.InterceptorPayload;
import com.tencent.redux.lifecycle.LifeCycleProxy;
import com.tencent.redux.middleware.Middleware;
import com.tencent.redux.middleware.Middleware.PageStateGetter;
import com.tencent.redux.middleware.MiddlewareUtil;
import com.tencent.redux.reducer.Reducer;
import com.tencent.redux.reducer.ReducerCollect;
import com.tencent.redux.reducer.ReducerUtils;
import com.tencent.redux.reducer.SubReducer;
import com.tencent.redux.util.IPlatform;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.concurrent.Future;

/**
 * Desc: Page = Feature + Effect + Reducer + middleware
 * Page最重要的功能：负责创建Store以及组织页面的Effect调度。
 *
 * 目的：将页面组装逻辑剥离到一个逻辑类，更方便的构建逻辑层测试。
 *
 * 为何要延迟加载Feature？
 * 这其实是有需求场景的，比如开发者需要等一些server的返回值才能决定目前的加载哪些组件，
 * 再比如：如果开发者需要定制组件的加载过程，也可以通过懒加载的方式进行。
 *
 * 针对组件间交互：
 * PageBus完全隔离掉子组件发送app级别的effect。
 *
 * @author randytu on 2021/6/27
 */
public abstract class LogicPage<S extends BasePageState> extends Logic<S> {
    /**
     * 当前App级别的Effect Bus
     */
    protected final DispatchBus mAppBus = ReduxManager.getInstance().getAppBus();

    /**
     * 页面依赖的Feature集合
     */
    protected DependentCollect<S> mDependencies;

    /**
     * Effect拦截器，用于实现子组件与子组件间的通信
     */
    protected Interceptor mInterceptor;

    /**
     * 用于取消异步任务
     */
    private Future<?> mFuture;

    /**
     * 构造器，初始Reducer/Effect以及一些依赖
     *
     * @param proxy LifeCycleProxy
     */
    public LogicPage(@NonNull LifeCycleProxy proxy) {
        super(proxy.getBundle());
        mEnvironment = Environment.of().setLifeCycleProxy(proxy);
        initDependencies();
        initInterceptor();
    }

    /**
     * 创建当前页面所用的Store，只在Page中创建页面Store
     */
    private void createStore(S state ) {
        // 1、将所有组件的reducer跟page Reducer合并
        Reducer<State> reducer = combineReducer();

        PageStore<BasePageState> store = new PageStore<>(reducer, state);
        mEnvironment.setStore(store);

        // 3、将Middleware链式化
        PageStateGetter getter = store::getAllState;
        List<Middleware> pageMiddleware = getReducerMiddleware();

        // 4、根据middleware生成最终的reducer
        final Dispatch dispatch = store.mDispatch;
        store.mDispatch =
                MiddlewareUtil.applyReducerMiddleware(pageMiddleware, dispatch, getter);
    }

    private void initDependencies() {
        // 获取Page对应的依赖集合
        if (mDependencies == null) {
            mDependencies = new DependentCollect<>();
        }
        addDependencies(mDependencies);
    }

    private void initInterceptor() {
        InterceptorCollect innerInterceptorCollect = new InterceptorCollect();

        HashMap<String, Dependant<? extends BaseComponentState, S>> map =
                mDependencies.getDependantMap();
        if (map == null || map.isEmpty()) {
            return;
        }

        for (Dependant<? extends BaseComponentState, S> dependant: map.values()) {
            LRConnector<? extends BaseComponentState, S> connector = dependant.getConnector();
            HashMap<Action, InterceptorBean> interceptors =
                    makeInterceptorForAction(connector, dependant);
            if (interceptors == null) {
                continue;
            }

            innerInterceptorCollect.add(interceptors);
        }

        mInterceptor = innerInterceptorCollect.getInterceptor();
    }

    private HashMap<Action, InterceptorBean> makeInterceptorForAction(
            LRConnector<? extends BaseComponentState, S> connector,
            Dependant<? extends BaseComponentState, S> dependant) {
        // 获取拦截器收集器
        HashMap<Action, ? extends Interceptor<? extends BaseComponentState>>
                interceptorCollect = connector.interceptorCollector();
        if (interceptorCollect == null || interceptorCollect.isEmpty()) {
            return null;
        }

        final HashMap<Action, InterceptorBean> interceptors = new HashMap<>();

        // 单一action可以被多个组件同时拦截，这个跟中间件是有区别的，
        // 中间件是一个全局性的操作，而拦截器是针对某个action的操作。
        // 这里创建Bean是为了让每个组件在拦截器里拿到的是自己对应的Context。
        for (Action action : interceptorCollect.keySet()) {
            InterceptorBean bean = new InterceptorBean();
            bean.mGetter = () -> dependant.getLogic().getContext();
            bean.mInterceptor = interceptorCollect.get(action);
            interceptors.put(action, bean);
        }

        return interceptors;
    }

    @SuppressWarnings("unchecked")
    private Reducer<State> combineReducer() {
        ArrayList<SubReducer> list = new ArrayList<>();

        // 合并所有组件的Reducer
        mergeReducer(list, null);

        // 合并sub reducer为Reducer
        Reducer childReducers = ReducerUtils.combineSubReducers(list);

        ArrayList<Reducer> reducers = new ArrayList<>();
        reducers.add(mReducer);
        reducers.add(childReducers);

        // 将所有的reducer合并成一个Reducer
        return ReducerUtils.combineReducers(reducers);
    }

    @Override
    final void mergeReducer(@NonNull List<SubReducer> list, LRConnector connector) {
        if (mDependencies != null) {
            mDependencies.mergerDependantReducer(list);
        }
    }

    @Override
    protected final void checkReducer(ReducerCollect<S> reducerCollect) {
        reducerCollect.remove(InnerActions.INTERCEPT_ACTION);
        reducerCollect.remove(InnerActions.INSTALL_EXTRA_FEATURE_ACTION);
        reducerCollect.remove(InnerActions.CHANGE_ORIENTATION);

        // 注册修改横竖屏状态的reducer
        reducerCollect.add(InnerActions.CHANGE_ORIENTATION,
                (getter, action, payload) -> {
                    S state = getter.copy();
                    state.mCurrentOrientation.set((int) payload);
                    return state;
                });
    }

    @Override
    protected final void checkEffect(EffectCollect<S> effectCollect) {
        effectCollect.remove(InnerActions.INTERCEPT_ACTION);
        effectCollect.remove(InnerActions.INSTALL_EXTRA_FEATURE_ACTION);
        effectCollect.remove(InnerActions.CHANGE_ORIENTATION);

        effectCollect.add(InnerActions.INTERCEPT_ACTION, makeInterceptEffect());
        effectCollect.add(
                InnerActions.INSTALL_EXTRA_FEATURE_ACTION,
                makeInstallExtraFeatureEffect());
    }

    @SuppressWarnings("unchecked")
    private Effect<S> makeInterceptEffect() {
        // 拦截Action的Effect
        return (action, ctx, payload) -> {
            if (!(payload instanceof InterceptorPayload)) {
                return;
            }

            InterceptorPayload interceptorPayload = (InterceptorPayload) payload;
            mInterceptor.doAction(interceptorPayload.mRealAction, ctx, payload);
        };
    }

    @SuppressWarnings("unchecked")
    private Effect<S> makeInstallExtraFeatureEffect() {
        // 安装额外的子组件的Effect
        return (action, ctx, payload) -> {
            if (!(payload instanceof ExtraDependants)) {
                return;
            }

            final HashMap<String, Dependant<? extends BaseComponentState, S>> extraFeatures =
                    ((ExtraDependants<S>)payload).mExtraFeature;
            if (extraFeatures == null || extraFeatures.isEmpty()) {
                return;
            }

            // 安装feature需要在主线程执行
            if (Looper.getMainLooper() == Looper.myLooper()) {
                installExtraDependant(extraFeatures);
                return;
            }

            ReduxManager.getInstance().submitInMainThread(() -> {
                installExtraDependant(extraFeatures);
            });
        };
    }

    private void installExtraDependant(HashMap<String,
            Dependant<? extends BaseComponentState, S>> extraDependants) {
        boolean hasNewFeature = false; // 防止多次调用重复安装

        // 当前Feature集合，这里保存的都是已经安装过的
        HashMap<String, Dependant<? extends BaseComponentState, S>> map =
                mDependencies.getDependantMap();

        // 子组件需要从父组件继承一些信息
        Environment env = Environment.copy(mEnvironment);
        env.setParentDispatch(env.getDispatchBus().getPageDispatch());
        env.setParentState(mEnvironment.getStore().mState);

        for (String key : extraDependants.keySet()) {
            if (map.containsKey(key)) {
                continue;
            }

            hasNewFeature = true;
            Dependant<? extends BaseComponentState, S> dependant = extraDependants.get(key);
            if (dependant != null) {
                map.put(key, dependant);
                // 安装子组件
                dependant.install(env);
            }
        }

        if (!hasNewFeature) {
            // 没有新的Feature被安装，直接返回
            return;
        }

        // 重新合并reducer
        mEnvironment.getStore().mReducer = combineReducer();

        // 重新收集拦截器
        initInterceptor();
    }

    /**
     * 创建Page bus，用于组件间Effect交互
     */
    private void createPageBus() {
        DispatchBus bus = new DispatchBus();
        bus.attach(mAppBus);
        mEnvironment.setDispatchBus(bus);
    }

    /**
     * 创建平台相关操作
     * @return IPlatform
     */
    protected IPlatform createPlatform() {
        return null;
    }

    /**
     *对于实际的Page组件，可能要设置额外的信息
     */
    protected void onStateDetected(S state) { }

    /**
     * 创建页面的ReduxContext，依赖一个初始的PureState
     */
    private void createContext() {
        if (ReduxManager.getInstance().getAsyncMode()) {
            createContextAsync();
        } else {
            createContextSync();
        }
    }

    private void createContextSync() {
        // 1、生成state
        final S state = onCreateState(mBundle);

        // 生成Key映射表
        state.detectField();

        // 负责处理额外的事情
        onStateDetected(state);

        // 2、创建Store
        createStore(state);

        // 3、创建Context
        mContext = new ReduxContextBuilder<S>()
                .setLogic(this)
                .setState(state)
                .setPlatform(createPlatform())
                .build();
        mContext.setController(getController());
        mContext.setStateReady();
    }

    /**
     * 异步模式，对性能极致需求
     */
    private void createContextAsync() {
        // 1、生成state
        final S state = onCreateState(mBundle);

        final Runnable detectFinishRunnable = () -> {
            mContext.setStateReady();
            // 负责处理额外的事情
            onStateDetected(state);
        };

        final Runnable detectRunnable = () -> {
            state.detectField();
            ReduxManager.getInstance().submitInMainThread(detectFinishRunnable);
        };

        // 整理state
        mFuture = ReduxManager.getInstance().submitInSubThread(detectRunnable);

        // 2、创建Store
        createStore(state);

        // 3、创建Context
        mContext = new ReduxContextBuilder<S>()
                .setLogic(this)
                .setState(state)
                .setPlatform(createPlatform())
                .build();
        mContext.setController(getController());
    }

    /**
     * Page收到容器onCreate生命周期时，执行的一些初始化操作
     */
    @CallSuper
    protected void onCreate() {
        // 1、创建Page bus，用于页面内Effect交互
        createPageBus();

        // 2、初始化Page的Context
        createContext();

        // 3、安装子组件
        installDependant();
    }

    /**
     * 安装子组件
     */
    private void installDependant() {
        HashMap<String, Dependant<? extends BaseComponentState, S>> map =
                mDependencies.getDependantMap();
        if (map == null || map.isEmpty()) {
            return;
        }

        // 子组件需要从父组件那边继承一些信息
        Environment env = Environment.copy(mEnvironment);
        env.setParentDispatch(env.getDispatchBus().getPageDispatch());
        env.setParentState(mEnvironment.getStore().mState);

        // 安装子组件
        for (Dependant<? extends BaseComponentState, S> dependant : map.values()) {
            dependant.install(env);
        }
    }

    /**
     * 创建页面的State，页面的State默认页面的一级组件都可以进行关联
     *
     * @return state {@link BasePageState}
     */
    public abstract S onCreateState(Bundle bundle);

    /**
     * 配置当前页面的Feature(依赖)集合PageDependantCollect，需要外部设置
     */
    public abstract void addDependencies(DependentCollect<S> collect);

    /**
     * 添加拦截 reducer 的 Middleware，开发这如果需要增加一些中间件拦截Action，可以通过此方法注入
     *
     * @return 中间件列表
     */
    protected List<Middleware> getReducerMiddleware() {
        // sub class can impl
        return null;
    }

    protected BaseController<S> getController() {
        return null;
    }

    @CallSuper
    protected void destroy() {
        if (mFuture != null && !mFuture.isDone()) {
            mFuture.cancel(true);
        }
    }
}

File: redux/src/main/java/com/cyworks/redux/component/LogicPage.kt
package com.cyworks.redux.component

import android.os.Bundle
import android.os.Looper
import androidx.annotation.CallSuper
import androidx.collection.ArrayMap
import com.cyworks.redux.DispatchBus
import com.cyworks.redux.ReduxContext
import com.cyworks.redux.ReduxContextBuilder
import com.cyworks.redux.ReduxManager
import com.cyworks.redux.action.Action
import com.cyworks.redux.action.InnerActionTypes
import com.cyworks.redux.dependant.Dependant
import com.cyworks.redux.dependant.DependentCollector
import com.cyworks.redux.dependant.ExtraDependants
import com.cyworks.redux.interceptor.InterceptorManager
import com.cyworks.redux.interceptor.InterceptorPayload
import com.cyworks.redux.lifecycle.LifeCycleProxy
import com.cyworks.redux.logic.EffectCollector
import com.cyworks.redux.state.ReflectTask
import com.cyworks.redux.state.ReflectTaskManager
import com.cyworks.redux.state.State
import com.cyworks.redux.state.StateType
import com.cyworks.redux.store.PageStore
import com.cyworks.redux.types.Effect
import com.cyworks.redux.types.Interceptor
import com.cyworks.redux.util.Environment
import com.cyworks.redux.util.ILogger
import com.cyworks.redux.util.IPlatform
import kotlin.reflect.full.memberProperties

/**
 * Page最重要的功能：负责创建Store以及组织页面的Effect调度。
 * 将页面组装逻辑剥离到一个逻辑类，更方便的构建逻辑层测试。
 *
 * 为何要延迟加载Dep
 * 这其实是有需求场景的，比如开发者需要等一些server的返回值才能决定目前的加载哪些组件，
 * 再比如：如果开发者需要定制组件的加载过程，也可以通过懒加载的方式进行。
 */
abstract class LogicPage<S : State>(p: Bundle?, proxy: LifeCycleProxy) : Logic<S>(p) {
    /**
     * 当前App级别的Effect Bus
     */
    private val appBus: DispatchBus = ReduxManager.instance.appBus

    /**
     * 拦截器，用于实现子组件与子组件间的通信
     */
    protected var interceptor: Interceptor<S>? = null

    private val interceptorManager: InterceptorManager = InterceptorManager()

    /**
     * 页面依赖的Feature集合
     */
    private var dependencies: DependentCollector<S>? = null

    override val childrenDepMap: ArrayMap<String, Dependant<out State, S>>?
        get() = if (dependencies == null) {
            null
        } else dependencies!!.dependantMap

    init {
        environment = Environment.of()
        environment.lifeCycleProxy  = proxy
        environment.taskManager = ReflectTaskManager()
        initDependencies()

        // 开始添加拦截器
        @Suppress("UNCHECKED_CAST")
        interceptor = interceptorManager.getInterceptor() as Interceptor<S>
        initInterceptor()
    }

    private fun initDependencies() {
        // 获取Page对应的依赖集合
        if (dependencies == null) {
            dependencies = DependentCollector()
        }
        addDependencies(dependencies)
    }

    private fun initInterceptor() {
        val map: ArrayMap<String, Dependant<out State, S>> = dependencies?.dependantMap ?: return
        val size = map.size

        for (i in 0 until size) {
            map.valueAt(i).mergeInterceptor(this.interceptorManager)
        }
    }

    /**
     * Page收到容器onCreate生命周期时，执行的一些初始化操作
     */
    @CallSuper
    protected open fun onCreate() {
        // 创建Page bus，用于页面内Effect交互
        createPageBus()

        // 初始化Page的Context
        createContext()

        // 安装子组件
        installSubComponents()
    }

    /**
     * 创建Page bus，用于组件间Effect交互
     */
    private fun createPageBus() {
        val bus = DispatchBus()
        bus.attach(appBus)
        environment.pageDispatchBus= bus
    }

    /**
     * 创建页面的ReduxContext，依赖一个初始的PureState
     */
    private fun createContext() {
        // 创建state
        val state = onCreateState(props)
        state.stateType = StateType.PAGE_TYPE

        // 反射解析属性
        reflect(state)

        // 创建当前页面所用的Store，只在Page中创建页面Store
        environment.store = PageStore(state)

        // 创建Context
        context = ReduxContextBuilder<S>()
            .setLogic(this)
            .setState(state)
            .setPlatform(createPlatform()!!)
            .build()
    }

    private fun reflect(state: S) {
        val rootTask = ReflectTask(1, environment.taskManager?.executor!!)

        val detectRunnable = Runnable {
            val time = System.currentTimeMillis()
            val memberList = state.javaClass.kotlin.memberProperties
            if (ReduxManager.instance.enableLog) {
                ReduxManager.instance.logger.d(ILogger.PERF_TAG,
                    "reflect ${state.javaClass.name} filed, consume: ${System.currentTimeMillis() - time}ms")
            }
            // 提交到主线程
            ReduxManager.instance.submitInMainThread {
                state.detectField(memberList)
                // 负责处理额外的事情
                onStateDetected(state)
                context.setStateReady()
                // 检查下一个任务
                environment.taskManager?.tryRunNextTask(rootTask, state.token)
                // 订阅属性
                logicModule.subscribeProps(context.state, propsWatcher)
            }
        }

        rootTask.add(state.token, detectRunnable)
        environment.taskManager?.putTask(rootTask)
    }

    /**
     * 安装子组件
     */
    private fun installSubComponents() {
        val map: ArrayMap<String, Dependant<out State, S>> = dependencies?.dependantMap ?: return
        val size = map.size
        if (size < 1) {
            return
        }

        // 子组件需要从父组件那边继承一些信息
        val env = copyEnvToChild()
        env.task = ReflectTask(map.size, environment.taskManager?.executor!!)

        // 安装子组件
        for (i in 0 until size) {
            map.valueAt(i).install(env)
        }
        environment.taskManager?.putTask(env.task!!)
    }

    protected abstract fun copyEnvToChild(): Environment

    final override fun checkEffect(effectCollector: EffectCollector<S>?) {
        effectCollector?.remove(InnerActionTypes.INTERCEPT_ACTION_TYPE)
        effectCollector?.remove(InnerActionTypes.INSTALL_EXTRA_FEATURE_ACTION_TYPE)
        effectCollector?.remove(InnerActionTypes.CHANGE_ORIENTATION_TYPE)

        effectCollector?.add(InnerActionTypes.INTERCEPT_ACTION_TYPE, makeInterceptEffect())
        effectCollector?.add(InnerActionTypes.INSTALL_EXTRA_FEATURE_ACTION_TYPE,
            makeInstallExtraFeatureEffect())
    }

    private fun makeInterceptEffect(): Effect<S> {
        // 拦截Action的Effect
        val effect: Effect<S> = Effect { action, ctx ->
            if (action.payload is InterceptorPayload) {
                interceptor?.doAction(action, ctx)
            }
        }

        return effect
    }

    private fun makeInstallExtraFeatureEffect(): Effect<S> {
        // 安装额外的子组件的Effect
        val effect: Effect<S> = object : Effect<S> {
            override fun doAction(action: Action<out Any>, ctx: ReduxContext<S>?) {
                if (action.payload !is ExtraDependants<*>) {
                    return
                }

                @Suppress("UNCHECKED_CAST")
                val extraFeatures: ArrayMap<String, Dependant<out State, S>>? =
                    (action.payload as ExtraDependants<S>).extra
                if (extraFeatures.isNullOrEmpty()) {
                    return
                }

                // 安装feature需要在主线程执行
                if (Looper.getMainLooper() == Looper.myLooper()) {
                    installExtraDependant(extraFeatures)
                    return
                }
                ReduxManager.instance.submitInMainThread { installExtraDependant(extraFeatures) }
            }
        }

        return effect
    }

    private fun installExtraDependant(extraDependants: ArrayMap<String, Dependant<out State, S>>) {
        var hasNewDep = false // 防止多次调用重复安装

        // 当前Feature集合，这里保存的都是已经安装过的
        val map: ArrayMap<String, Dependant<out State, S>> = dependencies?.dependantMap ?: return
        if (map.size < 1) {
            return
        }

        // 子组件需要从父组件继承一些信息
        val env = copyEnvToChild()
        val size = extraDependants.size
        env.task = ReflectTask(size, environment.taskManager?.executor!!)

        for (i in 0 until size) {
            val key = extraDependants.keyAt(i)
            if (map.containsKey(key)) {
                continue
            }

            hasNewDep = true
            val dependant: Dependant<out State, S>? = extraDependants[key]
            if (dependant != null) {
                map[key] = dependant
                // 安装子组件
                dependant.install(env)
            }
        }

        if (!hasNewDep) {
            // 没有新的Feature被安装，直接返回
            return
        }

        environment.taskManager?.putTask(env.task!!)
        // 重新收集拦截器
        initInterceptor()
    }

    /**
     * 创建平台相关操作
     * @return IPlatform
     */
    protected open fun createPlatform(): IPlatform? {
        return null
    }

    /**
     * 对于实际的Page组件，可能要设置额外的信息
     */
    protected open fun onStateDetected(state: S) {}

    /**
     * 创建页面的State，页面的State默认页面的一级组件都可以进行关联
     * @return state [State]
     */
    abstract fun onCreateState(props: Bundle?): S

    /**
     * 配置当前页面的Feature(依赖)集合PageDependantCollect，需要外部设置
     */
    abstract fun addDependencies(collect: DependentCollector<S>?)
}